File: hard/149. Max Points on a Line/149. Max Points on a Line.go
package main

import "fmt"

func gcd(a, b int) int {
	if b == 0 {
		return a
	}

	return gcd(b, a%b)
}

func maxPoints(points [][]int) int {
	res := 0
	n := len(points)

	for i := 0; i < n; i++ {
		hash := make(map[[2]int]int)
		maxCount := 0
		samePoint := 0
		x1, y1 := points[i][0], points[i][1]

		for j := 0; j < n; j++ {
			if i == j {
				samePoint += 1
				continue
			}

			x2, y2 := points[j][0], points[j][1]
			dx := x2 - x1
			dy := y2 - y1

			if dx == 0 && dy == 0 {
				samePoint += 1
			} else {
				g := gcd(dx, dy)
				dx /= g
				dy /= g

				if dx < 0 {
					dx = -dx
					dy = -dy
				}

				slope := [2]int{dy, dx}
				hash[slope] += 1

				if hash[slope] > maxCount {
					maxCount = hash[slope]
				}
			}
		}

		res = max(res, maxCount+samePoint)
	}

	return res
}

func main() {
	// [2]int{}
	fmt.Println(maxPoints([][]int{{1, 1}, {2, 2}, {3, 3}}))
	fmt.Println(maxPoints([][]int{{1, 1}, {3, 2}, {5, 3}, {4, 1}, {2, 3}, {1, 4}}))
}

File: hard/149. Max Points on a Line/149. Max Points on a Line.py
import math
from collections import defaultdict
from typing import List


class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        # Given an array of points where point = [x, y]
        # Use the line slope formula
        # (y2 - y1) / (x2 - x1)

        # For example: [[1, 4], [2, 3]]:
        # The slope is: (3 - 4) / (2 - 1) = -1 / 1 = -1
        # Next points: [[2, 3], [3, 2]]:
        # The slope is: (2 - 3) / (3 - 2) = -1 / 1 = -1
        # Next points: [[3, 2], [4, 1]]:
        # The slope is: (1 - 2) / (4 - 3) = -1 / 1 = -1
        # So points [[1, 4], [2, 3], [3, 2], [4, 1]] belong to the same line

        # {[slope]: count of points in line with this slope}
        n = len(points)
        res = 0

        for i in range(n):
            hash = defaultdict(int)
            x1, y1 = points[i][0], points[i][1]
            max_count = 0
            same_points = 0

            for j in range(n):
                if points[i] == points[j]:
                    same_points += 1
                    continue

                x2, y2 = points[j][0], points[j][1]
                dx = x2 - x1
                dy = y2 - y1

                if dx == 0:
                    slope = (float("inf"), 0)
                elif dy == 0:
                    slope = (0, float("inf"))
                else:
                    gcd = math.gcd(dx, dy)
                    slope = (dy // gcd, dx // gcd)

                hash[slope] += 1

                if hash[slope] > max_count:
                    max_count = hash[slope]

            line = max_count + same_points
            if line > res:
                res = line

        return res


solution = Solution()
print(solution.maxPoints(points=[[1, 1], [2, 2], [3, 3]]))  # 3
print(solution.maxPoints(points=[[1, 1], [3, 2], [5, 3], [4, 1], [2, 3], [1, 4]]))  # 4
print(
    solution.maxPoints(points=[[0, 0], [94911151, 94911150], [94911152, 94911151]])
)  # 1
print(solution.maxPoints(points=[[2, 3], [3, 3], [-5, 3]]))  # 3

# (1, -1): [
#   [[3, 2], [2, 3]],
#   [[3, 2], [1, 4]],
#   [[4, 1], [2, 3]],
#   [[4, 1], [1, 4]],
#   [[2, 3], [1, 4]]
# ]
